enhance: migrate cronjobs and scheduled tasks to new hatchet-based setup, improve scripting and cypress setup, migrate to vitest, dependency upgrades

sjschlapbach opened the pull request:
DEVELOPMENT BRANCH - DO NOT SQUASH ON MERGE

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Schedule and unpublish Live Quizzes from Manage (scheduling modal, UI labels, validation).

- **Improvements**
  - Centralized background scheduling/handlers for timed publish/end of quizzes, microlearnings, and activities.
  - Reverse proxy now exposes host port 8090; improved cross-platform startup/cleanup flow and more robust Redis-backed pub/sub and caching.

- **Tests**
  - GraphQL tests migrated to Vitest; Cypress split/parallel runs and a local test runner added.

- **Documentation**
  - Local testing guide and GraphQL README added.

- **Chores**
  - Tooling upgrades (Node/pnpm), dependency bumps, Docker build optimizations, CI workflow concurrency and caching refinements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sjschlapbach committed:
feat: set up self-hosted instance of hatchet (#4754)

sjschlapbach committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

sjschlapbach committed:
enhance: migrate cronjob-based activity publication and ending logic to hatchet scheduled tasks (#4756)

sjschlapbach committed:
enhance: migrate implemented cronjobs to be triggered through hatchet (#4793)

sjschlapbach committed:
feat: add possibilities to schedule draft live quizzes and unpublish scheduled live quizzes (#4794)

rschlaefli committed:
enhance: merge everything from v3, improve repo structure, simplify start scripts

rschlaefli committed:
fix: dual mode working, local CORS issue with https, move response api to new hatchet package, forward participant and temporary participant tokens from response api

rschlaefli commented on apps/hatchet-worker-general/src/index.ts:
```diff
@@ -1,13 +1,75 @@
 
 async function main() {
-  const workflows = prepareHatchetTasks(hatchet)
+  const redisExec = new Redis({
```
initialization should happen at the root (e.g., start of main function) -> as it is the same thing in backend, all workers, etc., we might also extract this into some kind of shared util function at some point..

rschlaefli commented on apps/hatchet-worker-general/src/index.ts:
```diff
@@ -1,13 +1,75 @@
-    // basic structure according to https://github.com/hatchet-dev/hatchet-typescript-quickstart/tree/main/monorepo
-    workflows,
+  const worker = await hatchetClient.worker('hatchet-worker-general', {
+    workflows: [
```
need to decide whether and how we would split this up into multiple workers. we could also do it "smartly" and pass the names of the workflows that should be loaded from the dict by a worker from an ENV variable. then we can configure the tasks in the deployment config and scale dynamically instead of hardcoding it.

rschlaefli replied:
implemented one approach to do so using an env variable like

HATCHET_WORKFLOWS=sendPushNotifications,...,...

rschlaefli commented on apps/backend-docker/src/index.ts:
```diff
@@ -6,6 +6,7 @@ import { withOptimize } from '@prisma/extension-optimize'
 // import '@sentry/tracing'
 import { createInMemoryCache, type Cache } from '@envelop/response-cache'
 import { createRedisCache } from '@envelop/response-cache-redis'
+import { hatchetClient, prepareHatchetTasks } from '@klicker-uzh/hatchet'
```
the hatchetClient that is imported here is already initialized with env variables etc. as a singleton pattern (similar to how prisma works when imported) -> we dont need to duplicate the hatchet client init stuff anymore

rschlaefli commented on apps/hatchet-worker-response-processor/src/index.ts:
```diff
@@ -34,9 +34,15 @@ export const processAuthenticatedResponseTask = hatchet.durableTask({
+    'hatchet-worker-response-processor',
+    {
+      workflows: [
+        processAuthenticatedResponseTask,
```
need to think whether temporary participants should be treated differently or not. not sure right now but these workflows will change a lot anyways

rschlaefli commented on apps/response-api/src/index.ts:
```diff
@@ -130,10 +130,24 @@ async function handleAddResponse(req: IncomingMessage, res: ServerResponse) {
-      ? req.headers['cookie']
-      : undefined
+  // Only forward participant-related cookies. If both exist, include both.
+  let cookie: string | undefined
```
we need to check that the manager cookie will not be forwarded with the message and that we are not counted as authenticated with the manage cookie but as anonymous. hence this "filtering" of cookies before forwarding. could certainly be done nicer, but the ai way works for now (and this entire thing will be rewritten in Go anyway)

rschlaefli commented on packages/graphql/src/lib/util.ts:
```diff
@@ -51,25 +50,6 @@ export function formatDate(dateTime: Date) {
   }
 }
 
-export async function sendTeamsNotifications(scope: string, text: string) {
```
moved to notifications as this is what it actually does

rschlaefli commented on packages/graphql/src/schema/practiceQuiz.ts:
```diff
@@ -176,6 +176,11 @@ export const ElementStack = ElementStackRef.implement({
 export interface IPracticeQuiz
   extends Pick<
     DB.PracticeQuiz,
+    | 'startedCount'
```
@sjschlapbach here I was not sure when merging, if this is wrong please correct. there was often a "three way merge" and it was a bit chaotic, but types work that way.

sjschlapbach replied:
fixed it :)

rschlaefli commented on packages/graphql/src/services/accounts.ts:
```diff
@@ -254,7 +254,7 @@ export async function sendMagicLink(
       variables: { LINK: magicLink },
     },
-    ctx
+    ctx.prisma
```
not sure if now it would actually be better to just import prisma everywhere and remove it from ctx altogether (now that it is a singleton in the prisma package). we can check and maybe do this separately though :)

rschlaefli commented on packages/graphql/src/services/notifications.ts:
```diff
@@ -148,58 +205,21 @@ export async function sendPushNotifications(_ctx: Context) {
+export async function sendTeamsNotifications(scope: string, text: string) {
+  if (process.env.TEAMS_WEBHOOK_URL) {
+    try {
+      return await axios.post(process.env.TEAMS_WEBHOOK_URL, {
```
TODO: migrate to new adaptive card mechanism (review how it was done in the course cms)

rschlaefli commented on packages/graphql/src/index.ts:
```diff
@@ -34,6 +34,17 @@ import './schema/subscription.js'
 //   })
 // }
 
+export {
```
these handlers are exported as they are needed in hatchet tasks. backend graphql or a worker will pass them down into the hatchet package. that way the two are completely decoupled (graphql is not installed in hatchet and also not the other way around).

rschlaefli commented on packages/hatchet/src/index.ts:
```diff
@@ -0,0 +1,481 @@
+import type { HatchetClient } from '@hatchet-dev/typescript-sdk'
```
we should probably split this into modules

rschlaefli commented on util/_create_hatchet_token.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
moved to util

rschlaefli commented on util/_create_ssl_certificates.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
```
created new script so we don't have to generate with mkcert anymore. automatically called on macos with the run app deps script

rschlaefli commented on _run_app_dependencies.sh:
```diff
@@ -1,7 +1,78 @@
 #!/bin/sh
```
consolidated all run scripts into one that detects the platform. to be tested on windows and linux but it looks much more sensible than separate scripts (especially now that we have additional logic that would also be repeated)

rschlaefli commented on package.json:
```diff
@@ -47,7 +47,7 @@
     "check:lint": "turbo run lint --parallel --continue",
     "check:syncpack": "syncpack lint",
-    "dev": "doppler run --config dev -- pnpm run dev:raw",
+    "dev": "CONFIG=dev ./util/_run_with_doppler.sh pnpm dev:raw",
```
will first use normal doppler if it works, but if not will fall back to the locally stored token (e.g., useful for external drives)

rschlaefli committed:
enhance: improve decoupling of graphql and hatchet, shared contracts in types, verified working e.g. scheduling LQ

rschlaefli commented on apps/backend-docker/src/index.ts:
```diff
@@ -1,11 +1,12 @@
 import { createRedisEventTarget } from '@graphql-yoga/redis-event-target'
-import { enhanceContext, schema } from '@klicker-uzh/graphql'
```
We export the "handlers" from Graphql -> functions that have to be available to any hatchet task from the graphql package. Reasoning -> we dont want to install graphql in hatchet and also not the other way around, as they are independent services and should be decoupled. The same way we provide hatchet tasks to graphql through the prepareHatchetTasks function and pass them down in context.

rschlaefli committed:
enhance: allow dynamic workflow selection in the general hatchet worker (allowing for separate deployment and scaling)

rschlaefli commented on packages/graphql/src/index.ts:
```diff
@@ -34,8 +36,28 @@ import './schema/subscription.js'
   },
 })
+
+export const handlers: HatchetHandlers = {
```
this will make sure that we also update the types whenever we add handlers

rschlaefli committed:
enhance: extract all hatchet handlers, begin improvements on graphql test suite matching new structure

chore(packages/graphql): rename handleSendTeamsNotification

fix: type and build issues due to hatchet updates

fix(ci): add cypress version of hatchet token preparation script

fix: cypress tests need different hatchet token generation approach (#4868)

fix: deployment config issues and dockerfile missing packages

chore: lockfile maintenance

deps: add hatchet sdk to dev dependencies of olat to fix issue

chore: add prisma migration, add prisma as dep to workers

chore: lockfile maintenance

deps: more peer dependencies for olat api

deps: more peer dependencies for olat api

chore: remove unneeded deps from olat-api

deps: updates

ci: extend check workflow

fix: add prisma to hatchet worker general dockerfile

fix: add types package to hatchet worker general dockerfile

rschlaefli committed:
fix: add all packages to both hatchet workers

rschlaefli commented:
<img width="1363" height="1239" alt="CleanShot 2025-09-01 at 21 30 15" src="https://github.com/user-attachments/assets/0efac6d0-a6b0-4b52-9e53-1147e8fabc58" />

<img width="1036" height="411" alt="CleanShot 2025-09-01 at 21 34 08" src="https://github.com/user-attachments/assets/d50fdbad-4e93-44f2-aafa-39b063931b52" />

Everything is now deployed to staging including the two workers. Many random issues with GH Actions/PNPM/Types already solved, but also many still remaining..

Response processing works with Hatchet but I believe the Functions are still broken.

rschlaefli commented on packages/transactional/out/MagicLinkRequested.html:
```diff
@@ -10,7 +10,7 @@
     * {
```
no idea why it reformatted these now

rschlaefli commented on .tool-versions:
```diff
@@ -1,2 +1,2 @@
-node 20.18.1
+node 20.19.4
 pnpm 9.8.0
```
```suggestion
pnpm 10.15.0
```

rschlaefli commented on docker-compose.build.yml:
```diff
@@ -0,0 +1,45 @@
+services:
```
to evaluate whether all containers build locally togehter with the util/_local_build.sh

rschlaefli committed:
refactor: migrate graphql tests to vitest, run containers only for services

rschlaefli committed:
fix(ci): update graphql test with separate hatchet database and new env variables

rschlaefli committed:
fix(ci): add hatchet variables to cypress test

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	apps/auth/package.json
#	pnpm-lock.yaml

rschlaefli committed:
fix(ci): build hatchet package in graphql test

rschlaefli committed:
chore: move grading package from jest to vitest for consistency

rschlaefli committed:
Merge branch 'v3' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	.github/workflows/check-types.yml
#	.github/workflows/v3_klickeruzhprod-response-processor(qa).yml
#	.github/workflows/v3_klickeruzhprod-response-processor-v2(qa).yml
#	.github/workflows/v3_klickeruzhprod-response-processor-v2.yml
#	.github/workflows/v3_klickeruzhprod-response-processor.yml
#	.github/workflows/v3_klickeruzhprod-responses(qa).yml
#	.github/workflows/v3_klickeruzhprod-responses.yml
#	apps/auth/package.json
#	apps/frontend-pwa/src/pages/session/[id].tsx
#	apps/func-incoming-responses/package.json
#	apps/func-response-processor/package.json
#	deploy/charts/klicker-uzh-v2/templates/cm-backend-graphql.yaml
#	packages/graphql/test/activityBatchOperations.test.ts
#	packages/graphql/test/helpers.ts
#	pnpm-lock.yaml

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	apps/auth/package.json
#	apps/backend-docker/package.json
#	apps/frontend-control/package.json
#	apps/frontend-pwa/package.json
#	apps/hatchet-worker-response-processor/Dockerfile
#	apps/hatchet-worker-response-processor/package.json
#	apps/lti/Dockerfile
#	apps/lti/package.json
#	package.json
#	packages/graphql/package.json
#	packages/util/package.json
#	pnpm-lock.yaml

rschlaefli committed:
fix(ci): check types for everything

rschlaefli committed:
refactor: separate prisma-data package, types as dependency, prevent cyclic-dep relations, add syncpack check

rschlaefli committed:
graphql test fixes

rschlaefli committed:
prisma before types build in gql test

rschlaefli committed:
fix graphql test workflow, naming of hatchet tasks, type issues in vitest

rschlaefli committed:
hatchet token in graphql test workflow

rschlaefli committed:
fix many type issues

rschlaefli committed:
fix script naming

rschlaefli committed:
calling shell, concurrency test

rschlaefli committed:
concurrency

rschlaefli committed:
fix type issue in ws server

sjschlapbach committed:
fix: resolve type errors caused by reintroduction of wrong interface types

sjschlapbach committed:
chore: resolve cosmetic issues in test files

sjschlapbach committed:
chore: add missing awaits

sjschlapbach committed:
chore: add proper publication and unpublishing tasks for all asynchronous activities

sjschlapbach committed:
chore: remove unnecessary prisma disconnects

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	apps/hatchet-worker-general/src/index.ts
#	apps/hatchet-worker-response-processor/package.json
#	apps/hatchet-worker-response-processor/src/index.ts
#	apps/response-api/package.json
#	apps/response-api/src/index.ts
#	docker-compose.yml
#	packages/hatchet-tasks/src/index.ts
#	pnpm-lock.yaml

rschlaefli committed:
valid mock for audit log

rschlaefli committed:
fix: make sure workers have a start:test command

rschlaefli committed:
fix: move defaults when undefined into Number conversion

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	apps/auth/package.json
#	apps/response-api/src/index.ts
#	packages/hatchet-tasks/src/index.ts
#	pnpm-lock.yaml

rschlaefli committed:
dependency updates/peers, hatchet worker general fixed with logging and workflow parsing

rschlaefli committed:
improve vitest script

rschlaefli committed:
remove prisma optimize integration for now, disable failing push notifications handler in hatchet task

rschlaefli committed:
update cypress tests

rschlaefli committed:
test improvements (CLEANUP at beginning, always end test when one fails, click instead of realClick where possible)

rschlaefli committed:
fixes with real clicks

rschlaefli committed:
fix tests, no experimental memory management

rschlaefli committed:
also start and configure response api for cypress

rschlaefli commented on cypress/cypress/e2e/J-elements-free-text-workflow.cy.ts:
```diff
@@ -1,33 +1,24 @@
+    }
+  })
+
+  it('CLEANUP', () => {
+    cy.cleanup()
+    cy.seed()
+  })
```
This does not work as Cypress runs the before hook also on origin change thus cleaning the DB after a part of the workflow when e.g. switching to the student app. So this change was on purpose.

rschlaefli committed:
selectOption, responses in live quizzes, Cypress issue

rschlaefli committed:
other way of closing dropdown

rschlaefli committed:
run on chrome and latest ubuntu

rschlaefli committed:
upload videos and screenshots to artifacts, keep for one day

rschlaefli committed:
127.0.0.1

rschlaefli committed:
more realClick for slate

rschlaefli committed:
try cypress split for draft PRs

rschlaefli committed:
improve if cases

rschlaefli committed:
improve conditions for jobs

rschlaefli committed:
improve triggers

rschlaefli committed:
further trigger improvements

rschlaefli committed:
10 containers in parallel

rschlaefli committed:
try with separate install and build step

rschlaefli committed:
change order of steps in build job

rschlaefli committed:
fail fast with variable and some realClick fixes

rschlaefli committed:
dont transfer node modules, reinstall in cypress

rschlaefli committed:
real clicks in template workflow

rschlaefli committed:
working template workflow

rschlaefli committed:
restore realClick in catalog workflow

rschlaefli committed:
one more realClick restored

rschlaefli committed:
restore resources workflow

rschlaefli committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

# Conflicts:
#	apps/hatchet-worker-response-processor/src/index.ts
#	packages/grading/package.json
#	pnpm-lock.yaml

rschlaefli committed:
build prisma in grading package

rschlaefli committed:
no separate build step for now

rschlaefli committed:
only run drafts on synchronize PR

rschlaefli committed:
Promise.all for cleanup

rschlaefli committed:
fix CYPRESS_FAIL_FAST variable

rschlaefli committed:
sequential cleanup

rschlaefli committed:
cleanup and seed in test:run (at the start of the runner)

rschlaefli committed:
change cookie domain for hatchet

sjschlapbach committed:
Merge branch 'v3-assessment' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

rschlaefli committed:
fix: create participations for assessment courses

rschlaefli committed:
real click, util peer dep

sjschlapbach committed:
chore: fix run app dependencies script

rschlaefli committed:
chore: add timings file and reduce testing library logging

rschlaefli committed:
Merge branch 'v3-hatchet' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

rschlaefli committed:
caching improvements in cypress

rschlaefli committed:
electron

rschlaefli committed:
old real clicks

rschlaefli committed:
restore more

rschlaefli committed:
fix: use realType whenever realClick is used to enter a field (e.g., for slate)

rschlaefli committed:
fix typo in realType

rschlaefli committed:
reorg project folder

rschlaefli committed:
enhance: pass execution context through to hatchet handlers (#4900)

rschlaefli committed:
try parallel for cy cloud as well

rschlaefli committed:
Merge branch 'v3-hatchet' of https://github.com/uzh-bf/klicker-uzh into v3-hatchet

rschlaefli committed:
service logs with matrix index

rschlaefli committed:
no parallel for now
